### PR TITLE
Can store your own request

### PR DESCRIPTION
**Pull request by Zemistr**

### CODE REVIEW

**Comment by Zemistr** [comment]
Now you can use something like this:

```php
<?php
namespace App\Modules\BackModule\Presenters;

use Nette\Application\UI\Presenter;

class Error4xxPresenter extends Presenter {
	public function renderDefault(\Exception $exception) {
		if ($exception->getCode() === 403) {
			$this->redirect('Sign:in', ['backlink' => $this->storeRequest('+ 30 minutes', $this->getParameter('request'))]);
		}

		// load template 403.latte or 404.latte or ... 4xx.latte
		$file = __DIR__ . "/templates/Error/{$exception->getCode()}.latte";
		$this->template->setFile(is_file($file) ? $file : __DIR__ . '/templates/Error/4xx.latte');
	}
}
```

**Comment by enumag** [comment]
This probably doesn't work for exceptions from createInitialRequest. I mean url `/admin/nonexistentpresenter` will trigger the global ErrorPresenter, not the one from AdminModule, right?

**Comment by Zemistr** [comment]
On row 172 - 174 is fallback ;)

**Comment by enumag** [comment]
Yes, that's what I'm talking about. I'd want the module specific error presenter to be triggered but instead the fallback will be used.

**Comment by dg** [comment]
ad 403: this indicates one of two conditions:

- Authentication was provided, but the authenticated user is not permitted to perform the requested operation.
- The operation is forbidden to all users. For example, requests for a directory listing return code 403 when directory listing has been disabled.

**Comment by Zemistr** [comment]
This change is doing simple thing "If exists error presenter in the same module, run him. If not, run global error presenter."

**Comment by Zemistr** [comment]
The example is just example @dg ;)

**Comment by enumag** [comment]
Yes, except that if a presenter does not exist but it is possible to determine module from routing it will be ignored.

**Comment by Zemistr** [comment]
Exists another solution for this thing?
I mean, can I set different error presenter for backend and for frontend?

**Comment by enumag** [comment]
@dg That's why you should extend ForbiddenRequestException and use instanceof to determine what to do with the exception instead of just error code. As an example this is sufficient.

**Comment by enumag** [comment]
@Zemistr You can look here for my solution but it's not good enough to put it to nette. https://github.com/enumag/Application/blob/master/src/Application.php

**Comment by Zemistr** [comment]
It seems unnecessarily complicated.
Why does my solution is not good?
I think that a modern application can not do without modules.

**Comment by enumag** [comment]
I know it's complicated, but it solves the problem I mentioned above.

**Comment by Zemistr** [comment]
I am little confused...what is wrong on my solution?
It works well with modules and works the same as is now in the stable release...

**Comment by dg** [comment]
It seems like unexpected convention. IMHO it can be done directly in ErrorPresenter, without need to modify Application.

**Comment by Zemistr** [comment]
Hmm...ok...and the second commit? :)

**Comment by enumag** [comment]
@Zemistr Your solution is good except for one case. Look at the createInitialRequest method. If the presenter is not found, an exception is [thrown](https://github.com/Zemistr/application/blob/master/src/Application/Application.php#L120) and the `$request` variable which holds the information about module is lost. In this case your solution will just use global error presenter (not because of your fallback but because `$this->presenter` is null and it falls to [this](https://github.com/Zemistr/application/blob/master/src/Application/Application.php#L179)). My solution preserves the request and uses it later to trigger the correct error presenter. This is why my solution is so complicated.

**Comment by Zemistr** [comment]
Okay, ignore the first commit. :)
What about the second commit?

**Comment by Zemistr** [comment]
Btw.:
This is my current global error presenter :)

```php
<?php
namespace App\Modules\Presenters;

use Nette;
use Nette\Application\Responses;
use Tracy\ILogger;

class ErrorPresenter extends Nette\Object implements Nette\Application\IPresenter {
	/** @var ILogger */
	private $logger;

	public function __construct(ILogger $logger) {
		$this->logger = $logger;
	}

	public function run(Nette\Application\Request $request) {
		$exception = $request->getParameter('exception');

		if ($exception instanceof Nette\Application\BadRequestException) {
			$presenter = explode(':', $request->getParameter('request')->getPresenterName(), 2);
			$presenter[1] = 'Error4xx';

			return new Responses\ForwardResponse($request->setPresenterName(implode(':', $presenter)));
		}

		$this->logger->log($exception, ILogger::EXCEPTION);

		return new Responses\CallbackResponse(
			function () {
				require __DIR__ . '/templates/Error/500.phtml';
			}
		);
	}
}

```

**Comment by enumag** [comment]
The second commit is good, I use it as well. I just have the parameters [swapped](https://github.com/enumag/Application/blob/master/src/UI/RequestStoragePresenterTrait.php#L39) with some additional magic so that I can call `$this->storeRequest($request);` without explicitly specifying expiration.

**Comment by TomasVotruba** [comment]
What needs to be done here to accepted or closed?

File: src/Application/UI/Presenter.php [hunk 1]
```diff
@@ -1082,17 +1082,18 @@ protected function handleInvalidLink(InvalidLinkException $e)
 
 	/**
 	 * Stores current request to session.
-	 * @param  mixed  optional expiration time
+	 * @param  mixed $expiration optional expiration time
+	 * @param  Application\Request $request optional specific request
 	 * @return string key
 	 */
-	public function storeRequest($expiration = '+ 10 minutes')
+	public function storeRequest($expiration = '+ 10 minutes', Application\Request $request = null)
```
Comment (dg) [inline src/Application/UI/Presenter.php]: should be `NULL`

**Comment by enumag** [comment]
As mentioned I'd prefer if the parameters were swapped. What do you think @dg?

File: src/Application/UI/Presenter.php [hunk 2]
```diff
@@ -1082,17 +1082,18 @@ protected function handleInvalidLink(InvalidLinkException $e)
 
 	/**
 	 * Stores current request to session.
-	 * @param  mixed  optional expiration time
+	 * @param  mixed $expiration optional expiration time
+	 * @param  Application\Request $request optional specific request
```
Comment (dg) [inline src/Application/UI/Presenter.php]: please remove `Application\Request $request`

File: src/Application/UI/Presenter.php [hunk 3]
```diff
@@ -1082,17 +1082,18 @@ protected function handleInvalidLink(InvalidLinkException $e)
 
 	/**
 	 * Stores current request to session.
-	 * @param  mixed  optional expiration time
+	 * @param  mixed $expiration optional expiration time
```
Comment (dg) [inline src/Application/UI/Presenter.php]: please remove `$expiration`

**Pull request closed without merging**